pandas/tests/arithmetic/test_object.py

```diff
@@ -323,3 +324,35 @@ def test_rsub_object(self):
 
         with pytest.raises(TypeError):
             np.array([True, pd.Timestamp.now()]) - index
+
+    def test_index_ops_defer_to_unknown_subclasses(self):
+        # https://github.com/pandas-dev/pandas/issues/31109
+        class MyIndex(pd.Index):
+
+            _calls: int
+
+            @classmethod
+            def _simple_new(cls, values, name=None, dtype=None):
+                result = object.__new__(cls)
+                result._data = values
+                result._index_data = values
+                result._name = name
+                result._calls = 0
+
+                return result._reset_identity()
+
+            def __add__(self, other):
+                self._calls += 1
+                return self._simple_new(self._index_data + other.to_list())
+
+            def __radd__(self, other):
+                return self.__add__(other)
+
+        values = np.array(
+            [datetime.date(2000, 1, 1), datetime.date(2000, 1, 2)], dtype=object
+        )
+        a = MyIndex._simple_new(values)
+        b = pd.timedelta_range("1D", periods=2, freq="D")
+        result = b + a
```

Comment:
Will this also dispatch correctly if `b` has another dtype? Eg numeric index? Or non-subclass object?

Reply:
It seems that numeric indexes don't go through this code path for their ops, so they don't dispatch. It's not clear to me that they should though... I've added those as xfails, but am OK with ignoring them.

Reply:
> It's not clear to me that they should though

Aside from the specific test case that brought this up, it would be very strange IMO that a datetime-like index dispatches and a numeric not?

Reply:
Perhaps. I'll need to investigate what the 0.25.3 behavior was. I don't think we should start returning NotImplemented where we didn't on 0.25.3, since subclasses may have been deferring to us.

Reply:
Until recently, the only object-dtype Index/ndarrays that we _didnt_ return NotImplemented for where those containing all-DateOffsets

Reply:
Here's the behavior for `Index.__add__(index_subclass)` from pandas 0.25.3 for

1. pandas index type
2. subclass dtype


self     | other   | result
-------- | ------- | ------
object   | object  | pandas
object   | numeric | pandas
numeric  | object  | pandas raises
numeric  | numeric | pandas raises
datelike | object  | NotImpemented
datelike | numeric | pandas

So while I'm not sure what the expected behavior is, I *think* this PR is restoring the 0.25 behavior... Except perhaps for object dtype index, which will now dispatch?

Reply:
@jorisvandenbossche do you have a preference here? Should we aim for restoring the 0.25.3 behavior (return NotImplemented only for `op(datetlike, object-dtype-subclass)`, or consistency (return NotImplemented for any subclass?)

My preference is probably for restoring the 0.25.3 behavior, which will require an adjustment to the PR to only return NotImplemented for datetlike `self`.

Reply:
> My preference is probably for restoring the 0.25.3 behavior, which will require an adjustment to the PR to only return NotImplemented for datetlike self.

I didn't really follow it implementation wise, but is it hard to go for consistency?

Reply:
Not hard, no. I would need to extend the subclass check to numeric indices. But it's a behavior change, and I could easily see an Index subclass intentionally deferring to Int64Index for ops. If so, they'd now get a TypeError since both side would return NotImplemented.

Reply:
OK, I am certainly fine with restoring the 0.25.3 behaviour.

But ideally, there is *some* logic in how the ops defer or not for the different classes we have ..